modules/gamAdServerSubmodule.js

```diff
@@ -0,0 +1,27 @@
+import { GAM_VENDOR } from './videoModule/constants/vendorCodes.js';
+import { adServerDirectory } from './videoModule/vendorDirectory.js';
```

Comment:
@dgirardi I'm a bit confused by the need to use `hook.js` since this does work as expected. At runtime the directory is updated by all registered modules, and the parent module is able to call on submodules accordingly. 
Why is `hook.js` required ? I prefer avoiding it because it seems like unnecessary code; it relies on a 3rd party package which seems like overkill for something this simple

Reply:
After pulling in the last update I did some digging and the reason it works is because you never actually read from `adServerDirectory` - if you did, you'd find it won't not contain what you expect. 

I'd like to give an example of what I mean (and why the rest of the codebase uses `hook.js`), but I am now confused as to the purpose of these registries. What do you intend them to be used for? can we sidestep the problem entirely by removing them for now? 

The general rule is modules should not import from other modules, only from core. The build process assumes that modules are independent; to simplify it, if you have two modules that do:

```javascript
// moduleA.js
export const someGlobalState = {};
export const someImmutableValue = 'foo';
```

```javascript
// moduleB.js
import {someGlobalState, someImmutableValue} from './moduleA.js';
someGlobalState[someImmutableValue] = 'bar';
```

they will get compiled to something like:

```javascript
symbols['moduleA'] = (function(){
    return {
       someGlobalState: {},
       someImmutableValue: 'foo'
    };
})();
symbols['moduleB'] = (function(){
   const {someGlobalState, someImmutableValue} = (function(){
    return {
       someGlobalState: {},
       someImmutableValue: 'foo'
    };
   })();
   moduleA[someImmutableValue] = 'bar';
   return {};
})();
```

that is, the build will not try to reuse the symbols exported from `moduleA`, because in general it cannot assume that it's  installed - you can choose to not include it in the package. Instead, it will re-compile it again and embed it in `moduleB`. If the symbols resolve to immutable values, this is just duplication - but if modules try to modify them, they will be each working on their copy, which will almost certainly be incorrect.

Reply:
Note that the above is the reason we have [this linter plugin](https://github.com/prebid/Prebid.js/blob/master/plugins/eslint/validateImports.js), which is supposed to error out when it detects cross-module imports. However, for reasons that I don't know, it allows cross-imports of anything except `index.js` if [the imported module is a directory](https://github.com/prebid/Prebid.js/blob/998e9343c472a1ed29037e7a6168a9771c5c13d4/plugins/eslint/validateImports.js#L25).

Reply:
The purpose of the `directories` is to serve as a registration process for submodules. When the prebid bundle is loaded on the page, the submodules included in the build will write to the shared directory. The `key` will be the submodule id string, and the `value` will be the submodule's factory function. When the parent module needs to instantiate a submodule, it can look up the factory. This way, a parent module does not need to import submodules; instead only the submodules included in the build will be linked to the parent module, by way of the `directory`

Reply:
the reason this works is because the `directory` is an `export const`

Reply:
no, the reason it works is because it doesn't do anything. At the moment the directories are useless, they are written to but never read.

Reply:
If it's useless then why does it work ? Have you tried the demo ? And actually the directory is being read
https://github.com/prebid/Prebid.js/pull/7264/files#diff-40c6bf504ebdf8d2f0211292ff3bc619bfca0448378f3bb0a0e658a463d37924R173

https://github.com/prebid/Prebid.js/pull/7264/files#diff-8f29ffeee22983e3a40ce2cb59a0deb6fb240a08bb8a993cc5fb50599a1f5664R75

Reply:
The demo, in the last version of this pr, throws an exception: 
> Uncaught TypeError: _videoModule_vendorDirectory_js__WEBPACK_IMPORTED_MODULE_2__.adServerDirectory is undefined

This is because [of this line](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/gamAdServerSubmodule.js#L2).

But that's just an oversight, and not the real issue. If you fix that and try to read from the ad server directory *from the video module*, you will see that it is empty. The steps I took are:

- change that import to use `videoVendorDirectory` instead of `adServerDirectory` (I know it's incorrect, but it does not matter, beacause imports like these do not work)
- change the video config to use `adServer: { provider: 'GAM' }`, so that you actually trigger [this line](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/videoModule/index.js#L110) to read from the directory
- put a breakpoint [here](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/shared/parentModule.js#L40) and watch `submodules` be empty.

Reply:
Sorry, my example was not a good one - you want the correct directory to be imported for it to be convincing (`import { adServerDirectory } from './adServerModule/vendorDirectory.js';`). But this affects every use of that paradigm if you cross module boundaries:

![no_submodules](https://user-images.githubusercontent.com/10151381/163866139-b66fde29-6f49-4e05-bccc-70c20bfca4b2.png)

Reply:
Ah I see, sorry I forgot to update the import statement in the gam submodule after moving the directory to the new adServer module folder. It should work now

Reply:
If you mean the demo, yes it does work, but the parentModule / directory pattern still doesn't - it's just not compatible with our module system. My suggestion is to either convert it to the module / submodule pattern defined in `hook.js`, or simplify this PR so that everything is contained in one module (at least to start; we can always extend later). The fact that the demo works shows that all these layers of abstractions are not needed right now - they are just being circumvented.

Also, I'm not sure if you see them on your end, but this PR has hundreds of LGTM alerts (https://lgtm.com/projects/g/prebid/Prebid.js/rev/pr-bf6d3f550989305375a32de43df4422b9c5c92b9), which makes it hard to follow - can you please address those?

Reply:
Yes regarding the LGTM alerts - by the way there's still one it's complaining about - unfortunately the bot is really keen on reminding us about them so it's drowning out real discussion in this thread.

Importing from modules results in incorrect code in the build output, as I tried to show above. All instances of `ParentModule` have an empty `submodules` because you cannot share state across modules this way. You would get the exact same behavior if you moved everything in `shared` and `adserver` inside the video module, because under the hood that's what webpack is doing. So that's at the minimum what's I'm asking you to do, because at least that way there's no extra dead code in the final bundle.

Reply:
Let me rephrase that - to a minimum, no module should `import` anything from any other module. It's unfortunate that our linter check happened to be blind to the way you are doing it here - I'm working on fixing that as well; I'm sorry you did all this work without knowing about this limitation.